x/bridge/observer/bitcoin/bitcoin.go

```diff
@@ -0,0 +1,301 @@
+func (b *ChainClient) fetchNewBlock() error {
+	nextHeight := b.lastObservedHeight.Load() + 1
+
+	// TODO: Now we infinitely retry to fetch one block if we get errors on either
+	// GetBlockHash or GetBlockVerboseTx. Decide what to do in that case.
+	// Skip the block after several retries?
+
+	hash, err := b.btcRpc.GetBlockHash(int64(nextHeight))
+	if err != nil {
+		var rpcErr *btcjson.RPCError
+		if errors.As(err, &rpcErr) && rpcErr.Code == btcjson.ErrRPCInvalidParameter {
+			// The block doesn't exist yet, skip it for now
+			return ErrBlockUnavailable
+		}
+		return errorsmod.Wrapf(err, "Failed to get block hash")
+	}
+
+	b.logger.With("height", nextHeight, "hash", hash).Info("Fetching bitcoin block")
+
+	blockVerbose, err := b.btcRpc.GetBlockVerboseTx(hash)
+	if err != nil {
+		return errorsmod.Wrapf(err, "Failed to get verbose block")
+	}
+
+	for _, tx := range blockVerbose.Tx {
+		txIn, isRelevant, err := b.processTx(uint64(blockVerbose.Height), &tx)
+		if err != nil {
+			b.logger.Error(fmt.Sprintf("Failed to process Tx %s: %s", tx.Txid, err.Error()))
+			continue
+		}
+		if !isRelevant {
+			continue
+		}
+
+		select {
+		case b.outboundChan <- txIn:
+		case <-b.stopChan:
+			b.logger.Info("Observer exiting early, tx skipped: ", tx.Txid)
+			return nil
+		}
+	}
+
+	b.lastObservedHeight.Add(1)
+	return nil
+}
+
+// processTx builds `Transfer` from provided Bitcoin transaction
+func (b *ChainClient) processTx(height uint64, tx *btcjson.TxRawResult) (observer.Transfer, bool, error) {
+	amount, relevant := b.isTxRelevant(tx)
+	if !relevant {
+		return observer.Transfer{}, false, nil
+	}
+
+	memo, contains := getMemo(tx)
+	if !contains {
+		return observer.Transfer{}, false, fmt.Errorf("failed to get Tx memo")
+	}
+
+	return observer.Transfer{
+		SrcChain: observer.ChainIdBitcoin,
+		DstChain: observer.ChainIdOsmosis,
+		Id:       tx.Hash,
+		Height:   height,
+		Sender:   "", // NB! the sender is set in the osmosis chain client
+		To:       memo,
+		Asset:    bridgetypes.DefaultBitcoinDenomName,
+		Amount:   amount,
+	}, true, nil
+}
+
+// isTxRelevant checks if a tx contains transfers to the BTC vault and calculates the total amount to transfer.
+// This method goes through all the Vouts and accumulates their values if Vouts.account[0] == vault.
+// Returns true is the tx is not relevant. Otherwise, false and the total amount to transfer.
+func (b *ChainClient) isTxRelevant(tx *btcjson.TxRawResult) (math.Uint, bool) {
+	var amount = sdk.NewUint(0)
+	for _, vout := range tx.Vout {
+		if strings.EqualFold(vout.ScriptPubKey.Type, "nulldata") {
+			continue
+		}
+		if vout.Value <= 0 {
+			continue
+		}
+
+		addresses, err := b.getAddressesFromScriptPubKey(vout.ScriptPubKey)
+		if err != nil || len(addresses) != 1 {
+			continue
+		}
+
+		if slices.Contains(addresses, b.vaultAddr) {
+			a, err := getAmount(vout)
+			if err != nil {
+				continue
+			}
+			amount = amount.Add(a)
+		}
```

Comment:
nit: if we only allow a len of 1 for addresses, we can just use the zero index instead of checking slice Contains, just to be extra explicit